### DISCUSSION TITLE
Entity Linking newbie

### DISCUSSION THREAD

**Discussion by surferzskan** [explosion/spaCy#8398] [Help: Coding & Implementations] [answered]
Hello, 
I'm new to the amazing Spacy ecosystem, it works great, but as a newbie, I'm finding some difficulties understanding the entity linking problem. If you have some time to help me understand it would be very helpful.
For what I understand @svlandeg is the domain expert here :)

I've corrected implemented a NER model with spacy v3.0. and it works well with my data. 
The next step is to try the entity linking. I was trying to follow the Emerson tutorial, but I'm finding it hard to relate that problem to my problem.

My Entities are not specific entities like the name of a person, but in my texts, I have _book numbers and their value_ (e.g. text = '.. the book n.2 has value 100 euros and ...'. In this text 'book n.2' is the entity 'Book',  and '100 euros' is the entity 'Value'; the link is from book to value).

**In this case, what the KnowledgeBase should be? Do I need it?**

I was trying to follow [this](![immagine](https://user-images.githubusercontent.com/84381244/122077978-831b7700-cdfc-11eb-8300-258f8e090d15.png)) tutorial, and it seems I don't need to create a KB. Is it true?

The second question is: 
**Which is the training data format for EL?**

In the NER, the training data were a list of tuples:
`[
     (text1, {'entities': [(start, end, 'LabelName'), (...)]}),
     (text2, {'entities': [(start, end, 'LabelName'), (...)]}),
     .....
]`

In the NEL problem, following the Emerson tutorial and the tutorial linked above, I'm building my training data as a list of Doc objects, where each Doc has the entities saved in doc.ents as Spans, and the relations I have built saved in the custom attribute doc._.rels in the format:

`{
     (start_parent_token, start_child_token) : {'LinkName' : 1.0, 'LinkName2': 0.0},
     (start_child_token, start_parent_token) : {'LinkName' : 0.0, 'LinkName2': 0.0},
     .....
}`

When I run the command **python -m spacy projects run train_cpu** it gives me the following error:
`ValueError: [E143] Labels for component 'relation_extractor' not initialized. This can be fixed by calling add_label, or by providing a representative batch of examples to the component's `initialize` method.`

What did I do wrong? What I'm missing?

Thanks for your help and for your amazing work.

S

**Comment by svlandeg** [comment, marked as answer]
Hi!

I could be mistaken, but I'm getting the feeling that you're confused with the terminology we've used for two distinct NLP challenges:

- "Entity Linking" (EL or NEL) is the challenge of standardizing (linking) an entity (mention in text) to an external identifier in a knowledge base. For example, the entity "Emerson" could be standardized/normalized/linked to [Q312545](https://www.wikidata.org/wiki/Q312545). While just the mention "Emerson" is an ambiguous piece of text, the unique ID Q312545 fully defines the entity in the "real world". 

- "Relation Extraction" (REL) is the challenge of linking two entities together because a certain relation exists between them - for example a relationship that says "Entity 1 regulates Entity 2", or "Entity 1 has the price of Entity 2". I assume this is what you're looking for in your use-case. If so, you might have better luck following [this tutorial](https://www.youtube.com/watch?v=8HL-Ap5_Axo), covered by [this example project](https://github.com/explosion/projects/tree/v3/tutorials/rel_component), and explained in more detail in the docs [here](https://spacy.io/usage/layers-architectures#component-rel).

With respect to setting the relations on your gold standard data, you can set the relations on the `Doc` directly (example code [here](https://github.com/explosion/projects/blob/v3/tutorials/rel_component/scripts/parse_data.py)).

**Reply by surferzskan** [reply]
Hi @svlandeg , thank you very much for your fast reply. 

Yes, I was confusing a little bit the two NLP challenges , as you correctly said what I'm looking for is the REL challenge.
My message was also not very clear, in fact, I was looking at the Emerson tutorial video to understand the Knowledge Base concept, but I cloned the rel_component example project and the tutorial, and I'm working on that. 

So my first question is now clear, I don't need the KB for REL.

The second question remains.

I have my training data as a list of Spacy Doc, where each Doc has the entities saved in doc.ents as Spans, and the relations saved in the custom attribute doc._.rel in the format:

`{ (start_parent_token, start_child_token) : {'LinkName' : 1.0, 'LinkName2': 0.0}, (start_child_token, start_parent_token) : {'LinkName' : 0.0, 'LinkName2': 0.0}, ..... }`

I didn't use Prodigy for the annotations, I had it in a file and I reconstructed them in the format above (dictionary with a tuple as key and another dict as value).

In the rel_component example tutorial, the only thing I changed in the project.yml is the path of the train, test, and dev data (vars section), which now point to my data (.spacy format).

When I run the command python -m spacy projects run train_cpu it gives me the following error:

> ValueError: [E143] Labels for component 'relation_extractor' not initialized. This can be fixed by calling add_label, or by providing a representative batch of examples to the component's initialize method.


Any help would be great.

Again, thank you very much,
Bests

S

EDIT: 
Searching for a solution, I tried 
`python -m debug config ./configs/rel_tok2vec.cfg`

and I have the following error:

> ValueError: [E002] Can't find factory for 'relation_extractor' for language English (en). This usually happens when spaCy calls `nlp.create_pipe` with a custom component name that's no
> t registered on the current language class. If you're using a Transformer, make sure to install 'spacy-transformers'. If you're using a custom component, make sure you've added the dec
> orator `@Language.component` (for function components) or `@Language.factory` (for class components).
> 
> Available factories: attribute_ruler, tok2vec, merge_noun_chunks, merge_entities, merge_subtokens, token_splitter, parser, beam_parser, entity_linker, ner, beam_ner, entity_ruler, lemm
> atizer, tagger, morphologizer, senter, sentencizer, textcat, textcat_multilabel, en.lemmatizer

Looking at [#7378 ](url), I noticed that in the custom_functions.py the imports you suggest in that discussion are already there:

```
# make the factory work
from scripts.rel_pipe import make_relation_extractor

# make the config work
from scripts.rel_model import create_relation_model, create_classification_layer, create_instances, create_tensors
```

Do I need them anywhere else?
Thank you again

EDIT 2:
Sorry for being such a nightmare. 
I tried the training pipeline using data from [this](url) tutorial and it works.
So the problem is probably in my training data, but I can't understand where.

**Reply by surferzskan** [reply]
Hi,

I managed to solve the problem.
During the saving on my training data in the DocBin format,
I wasn't using the parameter
`store_user_data=True`
so my relations were not being saved.

Now it seems to work.
Thank you very much for your help and your patience.

S

**Reply by svlandeg** [reply]
Happy to hear you got it resolved, and thanks for posting the solution! It might help others in their future search ;-)
